source/h2_stream.c

```diff
@@ -1335,14 +1338,16 @@ static int s_stream_write_data(
                 int error_code = stream->synced_data.api_state == AWS_H2_STREAM_API_STATE_INIT
                                      ? AWS_ERROR_HTTP_STREAM_NOT_ACTIVATED
                                      : AWS_ERROR_HTTP_STREAM_HAS_COMPLETED;
-                s_stream_data_write_destroy(stream, pending_write, error_code);
+                pending_write->error_code = error_code;
+                s_stream_data_write_destroy(stream, pending_write);
                 AWS_H2_STREAM_LOG(ERROR, stream, "Cannot write DATA frames to an inactive or closed stream");
                 return aws_raise_error(error_code);
             }
 
             if (stream->synced_data.manual_write_ended) {
                 s_unlock_synced_data(stream);
-                s_stream_data_write_destroy(stream, pending_write, AWS_ERROR_HTTP_MANUAL_WRITE_HAS_COMPLETED);
+                pending_write->error_code = AWS_ERROR_HTTP_MANUAL_WRITE_HAS_COMPLETED;
+                s_stream_data_write_destroy(stream, pending_write);
```

Comment:
debatable/trivial: it seems a bit fragile that we be so careful about setting write->error_code before calling destroy(). It would be easy to miss a spot, and have a write complete with error_code 0 in some weird error situation.

Maybe go back to passing error_code to the destroy() function.
If a data_frame fails due a write, then destroy the write immediately, instead of leaving it in a list to get cleaned up later. That's how it works with HTTP/1 chunked encoding ([see code](https://github.com/awslabs/aws-c-http/blob/ef78cb890afb9bf768afea6528b7f0fc324f361d/source/h1_encoder.c#L927-L938)). If failure was caused by the chunk, the chunk is immediately destroyed with that error code. Then connection shutdown continues...